libcudacxx/include/cuda/std/__type_traits/type_list.h

```diff
@@ -20,32 +21,807 @@
 #  pragma system_header
 #endif // no system header
 
-#include <cuda/std/cstddef>
+#include <cuda/std/__cccl/preprocessor.h>
+#include <cuda/std/__type_traits/conditional.h>
+#include <cuda/std/__type_traits/integral_constant.h>
+#include <cuda/std/__type_traits/is_same.h>
+#include <cuda/std/__type_traits/type_identity.h>
+#include <cuda/std/__type_traits/void_t.h>
+#include <cuda/std/__utility/integer_sequence.h>
+
+//! \file type_list.h
+//! This file defines a type-list type and some fundamental algorithms on type
+//! lists.
+//!
+//! It also defines a "meta-callable" protocol for parameterizing the type list
+//! algorithms, and some higher-order meta-callables. A meta-callable is a class
+//! type with a nested \c __call template alias.
+//!
+//! For the purpose of this file, a "trait type" is a class type with a nested
+//! type alias named \c type.
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-template <class _Hp, class _Tp>
-struct __type_list
+#ifndef DOXYGEN_SHOULD_SKIP_THIS // Do not document
+
+template <class... _Ts>
+struct __type_list;
+
+template <class _Tp>
+using __type = typename _Tp::type;
+
+//! \brief Evaluate a meta-callable with the given arguments
+template <class _Fn, class... _Ts>
+using __type_call = typename _Fn::template __call<_Ts...>;
+
+//! \brief Evaluate a unary meta-callable with the given argument
+template <class _Fn, class _Ty>
+using __type_call1 = typename _Fn::template __call<_Ty>;
+
+//! \brief Evaluate a binary meta-callable with the given arguments
+template <class _Fn, class _Ty, class _Uy>
+using __type_call2 = typename _Fn::template __call<_Ty, _Uy>;
+
+#  if defined(_CCCL_CUDACC_BELOW_12_2)
+
+namespace __detail
+{
+template <size_t _DependentValue>
+struct __type_call_indirect_fn
+{
+  template <template <class...> class _Fn, class... _Ts, class _Result = _Fn<_Ts...>>
+  _CCCL_HOST_DEVICE _LIBCUDACXX_HIDE_FROM_ABI static auto __fn(int) -> __type_identity<_Result>;
```

Comment:
`_LIBCUDACXX_HIDE_FROM_ABI` contains `_CCCL_HOST_DEVICE` So all of that should be
```suggestion
  _LIBCUDACXX_HIDE_FROM_ABI static auto __fn(int) -> __type_identity<_Result>;
```
Please apply throughout. If there is a compiler error we have other problems